operator/src/test/java/org/keycloak/operator/testsuite/apiserver/UpdateTest.java

```diff
@@ -0,0 +1,186 @@
+import org.keycloak.operator.controllers.KeycloakDeploymentDependentResource;
+import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
+import org.keycloak.operator.crds.v2alpha1.deployment.KeycloakStatusCondition;
+import org.keycloak.operator.crds.v2alpha1.deployment.ValueOrSecret;
+import org.keycloak.operator.crds.v2alpha1.deployment.spec.UpdateSpec;
+import org.keycloak.operator.testsuite.integration.BaseOperatorTest;
+import org.keycloak.operator.testsuite.utils.CRAssert;
+import org.keycloak.operator.update.UpdateStrategy;
+
+import io.fabric8.kubeapitest.KubeAPIServer;
+import io.fabric8.kubernetes.api.model.apps.StatefulSet;
+import io.fabric8.kubernetes.client.Config;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientBuilder;
+import io.fabric8.kubernetes.client.informers.ResourceEventHandler;
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+public class UpdateTest {
+
+    private static KubeAPIServer kubeApi;
+    static KubernetesClient k8sclient;
+
+    @BeforeAll
+    static void beforeAll() throws FileNotFoundException {
+        kubeApi = new KubeAPIServer();
+        kubeApi.start();
+        k8sclient = new KubernetesClientBuilder().withConfig(Config.fromKubeconfig(kubeApi.getKubeConfigYaml()))
+                .build();
+        BaseOperatorTest.createCRDs(k8sclient);
+    }
+
+    @AfterAll
+    static void afterAll() {
+        kubeApi.stop();
+    }
+
+    @Test
+    public void testExplicitStrategy() throws InterruptedException, ExecutionException, TimeoutException {
+        var operator = BaseOperatorTest.createOperator(k8sclient);
+        operator.start();
+
+        var kc = createInitialDeployment(UpdateStrategy.EXPLICIT);
+
+        var updateCondition = assertUnknownUpdateTypeStatus(kc);
+
+        // fake statefulset controller
+        k8sclient.apps().statefulSets().withName(KeycloakDeploymentDependentResource.getName(kc))
+                .inform(new ResourceEventHandler<StatefulSet>() {
+
+            @Override
+            public void onAdd(StatefulSet obj) {
+                updateStatefulSet(obj);
+            }
+
+            private void updateStatefulSet(StatefulSet obj) {
+                int replicas = obj.getSpec().getReplicas();
+                String revision = String.valueOf(Math.abs(obj.getSpec().hashCode()));
+                if (!Objects.equals(Optional.ofNullable(obj.getStatus().getReplicas()).orElse(0), replicas)
+                        || !Objects.equals(revision, obj.getStatus().getUpdateRevision())
+                        || !Objects.equals(obj.getStatus().getCurrentRevision(), obj.getStatus().getUpdateRevision())) {
+                    // generate intermediate rolling events
+                    int actualReplicas;
+                    if (!Objects.equals(revision, obj.getStatus().getUpdateRevision())) {
+                        actualReplicas = 0;
+                    } else if (replicas == 0) {
```

Comment:
Can this branch be removed? 
If the replicas < status.replicas then you decrement the number of replicas, otherwise you increment.

Reply:
If the revision doesn't match the updateRevision, then we are rolling in another spec change. Setting the replicas to 0 represents taking everything offline, then we let subsequent events handle the roll up. Hybrid of serial and parallel handling.

You don't have to have a replica change with a spec change, so the other handling with incrementing / decrementing won't do anything if the replicas are left the same in the CR.

Reply:
Right! forgot about the rolling update :facepalm: